File: benchmarks/cross/cases/data/byte_builder.py
import sys


APPENDS_PER_REPETITION = 80_000
OUTPUT_SIZE = 800_000
CHECKSUM_MULTIPLIER = 1_000_003
CHECKSUM_MODULUS = 1_000_000_007
CHUNK = b"0123456789"


def requested_work() -> int:
    if len(sys.argv) < 2:
        return 1
    try:
        return max(0, int(sys.argv[1]))
    except ValueError:
        return 1


def one_repetition() -> int:
    output = bytearray()
    for _ in range(APPENDS_PER_REPETITION):
        output.extend(CHUNK)
    return len(output) + (0 if not output else 1)


def main() -> None:
    checksum = 0
    for _ in range(requested_work()):
        checksum = (
            checksum * CHECKSUM_MULTIPLIER + one_repetition()
        ) % CHECKSUM_MODULUS
    print(checksum)


if __name__ == "__main__":
    main()

File: benchmarks/cross/cases/data/byte_builder.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

const (
	appendsPerRepetition = 80000
	outputSize           = 800000
	checksumMultiplier   = uint64(1000003)
	checksumModulus      = uint64(1000000007)
)

var chunk = []byte("0123456789")

func requestedWork() uint64 {
	if len(os.Args) < 2 {
		return 1
	}
	work, err := strconv.ParseInt(os.Args[1], 10, 64)
	if err != nil {
		return 1
	}
	if work < 0 {
		return 0
	}
	return uint64(work)
}

func oneRepetition() uint64 {
	output := make([]byte, 0, outputSize)
	for i := 0; i < appendsPerRepetition; i++ {
		output = append(output, chunk...)
	}
	if len(output) == 0 {
		return uint64(len(output))
	}
	return uint64(len(output) + 1)
}

func main() {
	work := requestedWork()
	var checksum uint64
	for rep := uint64(0); rep < work; rep++ {
		checksum = (checksum*checksumMultiplier + oneRepetition()) % checksumModulus
	}
	fmt.Println(checksum)
}
